Corrected Radio.h & Radio.m

DougPA opened the pull request:
Uncommented UIKit references (could these be surrounded by #if TARGET_OS_IPHONE  #endif)

Removed ResponseCallbacks for Global & TX Profiles

Added parseProfileToken method

Did not change profile arrays to dictionaries (can do in the future if desired)

Commit:
Initial commit of changes to make smartsdr compile as Framework

Commit:
Additions to Radio.m & Radio.h to implement Global & Tx Profiles

Commit:
Uncommented UIKit references

Removed ResponseCallbacks for Global & Tx Profiles

Added parseprofileToken method

Commit:
Add version, ant list & mic list commands in initializeRadio

Add ResponseCallbacks for version, ant list & mic list

add properties to hold values returned by each callback

Commit:
Bracketed #import <UIKit/UIKit.h> with #if OS_IS_IPHONE .... #endif (panafall.h & Radio.h)

Added declarations for cmdSaveGlobalProfile, cmdDeleteGlobalProfile, cmdSaveTxProfile & cmdDeleteTxProfile (Radio.h)

K6TU commented on Radio.m:
```diff
@@ -1445,14 +1577,62 @@ - (void) parseGpsToken: (NSScanner *) scan {
+    // List or Current value?
+    if ([profileSubType isEqualToString: @"list"]) {
+        // it's the List, separate the components
+        NSMutableArray *profileNames;
```
Good for these to be mutable as they will have profiles added/removed dynamically

K6TU commented on Radio.m:
```diff
@@ -986,6 +993,131 @@ - (void) infoResponseCallback:(NSString *)cmdResponse {
 
 
+/**
+ * Process a response from a Version command
```
This is really picky - sorry, but please replace this style of comment with

//
// Comment
//

And not use /\* */ comment delimiters - this is for consistency with all the other comments

K6TU commented on Radio.m:
```diff
@@ -1445,14 +1577,62 @@ - (void) parseGpsToken: (NSScanner *) scan {
+ *     scan is initially at scanLocation = 17, start of the <profileType>
+ *     "<apiHandle>|profile " has already been processed
+ */
 - (void) parseProfileToken: (NSScanner *) scan {
```
What happens when a profile is deleted from the radio?  Does the radio tell you the profile (by name) has been removed or does it just send the entire new list?

Also, with objects like the list of profile names that can be accessed on multiple threads, its critical to guard changes to the Array property in a @synchronized block - like

@synchronized(self.txProfiles) {
    // Add or remove object
}

How does the array remain consistent when possibly multiple clients are adding/removing profiles and we are getting status updates from the radio?

Commit:
Changed comment style
Removed //FIXME:
Made antList & micList arrays NOT mutable
Used updateWithNotify in antList, micList & version
Added @synchronized to txProfiles & globalProfiles update
Made antList, micList & version properties readonly (outside the class)

K6TU commented on Radio.m:
```diff
@@ -986,6 +1000,138 @@ - (void) infoResponseCallback:(NSString *)cmdResponse {
+    
+    // Anything other than 0 is an error
+    if ([errorNumAsString intValue] != 0) {
+        // FIXME: Do something?
```
Please update the comment to - ignore or something else.

K6TU commented on Radio.m:
```diff
@@ -1445,14 +1591,66 @@ - (void) parseGpsToken: (NSScanner *) scan {
+        // save it in the appropriate property
+        if ([profileType isEqualToString: @"global"]) {
+            @synchronized (self.globalProfiles) {
+                _globalProfiles = profileNames;
```
this needs to be an updateAndNotify so that a client of the model can KVO for changes on the array.

Although Flex hasn't implemented profiles this way, I think from a clients point of view there should be a ProfileCreated and a ProfileDeleted notification.  Its possible that someone may be KVO'ing the array and if so, they need to be notified before the array is changed so that the observer can de-register or else an exception will be thrown.  

I'm surprised that they didn't add a status notification telling a client that a profile has been removed rather than just dumping out a list of all the names.

Also, I notice that flex uses the caret ^ to replace spaces in profile names.  That should be handled transparently to our client here.  Both for setting the names in the arrays plus for removing spaces when the profile name is sent to the radio.

DougPA replied:
> On Jul 8, 2015, at 7:18 PM, Stu Phillips notifications@github.com wrote:
> 
> Also, I notice that flex uses the caret ^ to replace spaces in profile names. That should be handled transparently to our client here. Both for setting the names in the arrays plus for removing spaces when the profile name is sent to the radio

The caret isn’t replacing spaces, it’s used as the delimiter between Profile Names (which may contain spaces, e.g. “Default FHM-1 DX”). A typical profile status string looks like this:

```
S78613D02|profile tx list=Default^Default FHM-1^Default FHM-1 DX^Default HM-Pro^Default PR781^. . . ^RadioSport WIDE M360/EM56^
```

I can see why they did this, it allows you to have Profile Names with embedded spaces (but no carets). As long as we send the messages back to the Radio with caret delimiters we should be good.

A similar approach was taken in the versionResponseCallback message where they used hash (“#”) as the delimiter between “key=value” pairs. A typical one of those looks like this:

```
R13|0|SmartSDR-MB=1.4.11.60#PSoC-MBTRX=1.3.1.0#PSoC-MBPA100=1.1.2.2#FPGA-MB=0.0.23.167
```

I’m not sure why they did it here but the good news is we never send a message like this to the Radio.

73’s 
Doug - K3TZR

DougPA replied:
In the previous reply I said "As long as we send the messages back to the Radio with caret delimiters we should be good."

I was wrong here, we don't have a command to send a list of Profiles so we don't ever construct a list delimited by caret.

DougPA replied:
I did find a curious item in the C# code. Whenever they Save or Delete a TxProfile they replace all "*" with "" as in:

[profile stringByReplacingOccurrencesOfString:@"*" withString:@""]

I've added this into the code.

DougPA replied:
Stu said "there should be a ProfileCreated and a ProfileDeleted notification"

In the cmd... methods (cmdSaveGlobalProfile, cmdDeleteGlobalProfile, cmdSaveTxProfile, cmdDeleteTxProfile) there are currently "GlobalProfileCreated", "GlobalProfileUpdated", "GlobalProfileDeleted" notifications being sent. Similarly there are "TxProfileCreated", "TxProfileUpdated", "TxProfileDeleted" notifications being sent

K6TU commented on Radio.m:
```diff
@@ -1445,14 +1591,66 @@ - (void) parseGpsToken: (NSScanner *) scan {
+        // it's the Current value
+        // save it in the appropriate property
+        if ([profileType isEqualToString: @"global"]) {
+            _currentGlobalProfile = remainderOfCommand;
```
updateAndNotify required...

K6TU commented on Radio.m:
```diff
@@ -1445,14 +1591,66 @@ - (void) parseGpsToken: (NSScanner *) scan {
+        if ([profileType isEqualToString: @"global"]) {
+            _currentGlobalProfile = remainderOfCommand;
+        } else if ([profileType isEqualToString: @"tx"]) {
+            _currentTxProfile = remainderOfCommand;
```
updateAndNotify required

Commit:
When setting _globalProfiles, _txProfiles, _currentGlobalProfile or _currentTxProfile now use updateWithNotify
